### PR TITLE
AS Router Endpoint

### PR DESCRIPTION
**Pull request by glogiotatidis**
Implements a new endpoint `/6/` that serves (brotli) compressed json bundles for the new Activity Stream Router.

Interesting bits:
- Creates a new entry `/admin` named `Snippets NG` (Snippets Next Generation) but continues to use the `Snippets` model.
-  A new Form (`SnippetNGAdminForm`) for Admin displays only the currently available filters for snippets. 
- Implemented `fluent_link_extractor` to get links off `text` and into a separate list which is a fluent.js compatible format.
- Added Template `code_name` and `version`, also a simple filter to only list supported templates based on startpage version while editing SnippetNG and SnippetNG snippets.

### CODE REVIEW

File: snippets/base/admin.py [hunk 1]
```diff
@@ -265,9 +266,62 @@ def preview_url(self, obj):
...
-    def queryset(self, request):
-        return (super(SnippetAdmin, self)
-                .queryset(request).prefetch_related('locales').select_related('template'))
+    def get_queryset(self, request, filtr=True):
+        query = super(SnippetAdmin, self).get_queryset(request)
+        # When filtr it means that this is coming form the SnippetNG admin and
+        # will be filtered later to include only stage page 6 snippets.
+        if filtr:
+            query = query.exclude(on_startpage_6=True)
+        return query.prefetch_related('locales')
```
Comment (jpetto) [inline snippets/base/admin.py]: Do we no longer need the `select_related('template')` optimization?

File: snippets/base/admin.py [hunk 2]
```diff
@@ -265,9 +266,62 @@ def preview_url(self, obj):
...
+
+    fieldsets = (
+        (None, {'fields': ('name', 'published', 'campaign', 'preview_url', 'created', 'modified')}),
+        ('Content', {
+            'fields': ('template', 'data'),
+        }),
+        ('Publish Duration', {
+            'description': ('When will this snippet be available? (Optional)'
+                            '<br>Publish times are in UTC. '
+                            '<a href="http://time.is/UTC" target="_blank">'
```
[the remark this answers is not in the document]
Comment (pmac) [reply snippets/base/admin.py]: Great catch!

File: snippets/base/models.py [hunk 3]
```diff
@@ -234,22 +187,28 @@ def snippets(self):
...
+                'settings': settings,
+                'current_firefox_major_version': util.current_firefox_major_version(),
+            })
 
         if isinstance(bundle_content, unicode):
             bundle_content = bundle_content.encode('utf-8')
 
-        if settings.BUNDLE_BROTLI_COMPRESS and self.client.startpage_version == '5':
+        if ((settings.BUNDLE_BROTLI_COMPRESS and
+             (self.client.startpage_version == '5' or self.client.startpage_version == '6'))):
```
[the remark this answers is not in the document]
Comment (glogiotatidis) [reply snippets/base/models.py]: Fixed

File: snippets/base/admin.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (glogiotatidis) [reply snippets/base/admin.py]: I didn't know about this, thanks!

[the remark this answers is not in the document]
Comment (glogiotatidis) [reply snippets/base/admin.py]: Fixed here and in other places where we link to third parties and use `_blank`

File: snippets/base/models.py [hunk 4]
```diff
@@ -67,7 +64,7 @@
     )).hexdigest()
 
 CHANNELS = ('release', 'beta', 'aurora', 'nightly', 'esr')
-FIREFOX_STARTPAGE_VERSIONS = ('1', '2', '3', '4', '5')
+FIREFOX_STARTPAGE_VERSIONS = ('1', '2', '3', '4', '5', '6')
```
[the remark this answers is not in the document]
Comment (glogiotatidis) [reply snippets/base/models.py]: Fair enough, added comments!

File: snippets/base/forms.py [hunk 5]
```diff
@@ -177,9 +178,38 @@ def save(self, *args, **kwargs):
...
+                  'on_release', 'on_beta', 'on_aurora', 'on_nightly', 'on_esr',
+                  'on_startpage_6', 'campaign')
+        widgets = {
+            'data': TemplateDataWidget('template'),
+        }
+
+    def clean(self):
+        cleaned_data = super(SnippetNGAdminForm, self).clean()
+
+        validate_as_router_fluent_variables(cleaned_data['data'])
```
[the remark this answers is not in the document]
Comment (glogiotatidis) [reply snippets/base/forms.py]: Indeed your guess is correct. `text` and also other template fields are populated dynamically using JS and we currently don't have data on which is required or not. Filed issue #606 to implement the automatic template import from AS and added a note to implement to mark fields as required as the [new template schema](https://github.com/mozilla/activity-stream/blob/master/content-src/asrouter/templates/SimpleSnippet/SimpleSnippet.schema.json#L87) provides this info.

For now I made the validator to return if `text` is not present.

File: snippets/base/admin.py (hunk 1 above)
Comment (glogiotatidis) [reply snippets/base/admin.py]: This was a optimization put in place back in the day to tackle some SCL3 -> PHX1 connection timeouts. This is not needed anymore and either way it wasn't used b/c `queryset` has been replaced by `get_queryset` a couple Django versions back.

Good point though, thanks!

**Comment by glogiotatidis** [comment]
@jpetto updated!

**Pull request merged**